ui-lib/src/main/java/co/electriccoin/zcash/ui/common/FlowExt.kt

```diff
@@ -0,0 +1,59 @@
+package co.electriccoin.zcash.ui.common
+
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import kotlin.time.Duration
+import kotlin.time.ExperimentalTime
+import kotlin.time.TimeSource
+
+@OptIn(ExperimentalTime::class)
+fun <T> Flow<T>.throttle(
+    duration: Duration,
+    timeSource: TimeSource = TimeSource.Monotonic
+): Flow<T> = flow {
+    coroutineScope {
+        val context = coroutineContext
+        val mutex = Mutex()
+
+        var timeMark = timeSource.markNow()
+        var delayEmit: Deferred<Unit>? = null
+        var firstValue = true
+        var valueToEmit: T
```

Comment:
@ccjernigan something like this?

Reply:
Yes. But I think you still have a race condition.  valueToEmit gets assigned and then you call `delayEmit?.cancel()`.  Cancellation needs to occur first, otherwise it can emit the value twice in a row.